src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,275 @@
+ */
+
+import {layoutRectLtwh, rectIntersection, moveLayoutRect} from './layout-rect';
+
+const DEFAULT_THRESHOLD = [0, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4,
+    0.45, 0.5, 0.55, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 1];
+
+/**
+ * The structure that defines the rectangle used in intersection observers.
+ *
+ * @typedef {{
+ *   top: number,
+ *   bottom: number,
+ *   left: number,
+ *   right: number,
+ *   width: number,
+ *   height: number,
+ *   x: number,
+ *   y: number,
+ * }}
+ */
+export let DOMRect;
+
+/**
+ * Transforms a LayoutRect into a DOMRect for use in intersection observers.
+ * @param {!./layout-rect.LayoutRectDef} rect
+ * @return {!DOMRect}
+ */
+function DomRectFromLayoutRect(rect) {
+  return {
+    left: rect.left,
+    top: rect.top,
+    width: rect.width,
+    height: rect.height,
+    bottom: rect.bottom,
+    right: rect.right,
+    x: rect.left,
+    y: rect.top,
+  };
+}
+
+/**
+ * Transforms a DOMRect into a LayoutRect for use in intersection observers.
+ * @param {!DOMRect} rect
+ * @return {!./layout-rect.LayoutRectDef}
+ */
+function LayoutRectFromDomRect(rect) {
+  return {
+    left: rect.left,
+    top: rect.top,
+    width: rect.width,
+    height: rect.height,
+    bottom: rect.bottom,
+    right: rect.right,
+  };
+}
+
+
+/**
+ * Returns the ratio of the smaller box's area to the larger box's area.
+ * @param {!./layout-rect.LayoutRectDef} smaller
+ * @param {!./layout-rect.LayoutRectDef} larger
+ * @return {number}
+ */
+function intersectionRatio(smaller, larger) {
+  return (smaller.width * smaller.height) / (larger.width * larger.height);
+}
+
+/**
+ * The IntersectionObserverPolyfill class lets any element receive its
+ * intersection data with the viewport. The IntersectionObserverPolyfill acts
+ * like native browser supported IntersectionObserver.
+ * IntersectionObserverPolyfill will
+ * create a PositionObserver class (Not implemented yet) to track the
+ * intersection between element and viewport.
+ * The IntersectionObserver receives a callback function and an optional option
+ * as params. If option is not provided, the default thresholds value will be
+ * used. Whenever the element intersection ratio cross a threshold value,
+ * IntersectionObserverPolyfill will call the callback function.
+ * Note: To share the IntersectionObserverEntry to an iframe, whoever create the
+ * IntersectionObserverPolyfill needs to handle postMessageApi and takes care of
+ * sending messages rate etc outside the IntersectionObserverPolyfill.
+ */
+export class IntersectionObserverPolyfill {
+  /**
+   * @param {!function()} callback.
+   * @param {Object=} opt_option
+   */
+  constructor(callback, opt_option) {
+    /** @private @const {function()} */
+    this.callback_ = callback;
+
+    /**
+     * A list of threshold, sorted in increasing numeric order
+     * @private @const {!Array}
+     */
+    this.threshold_ = opt_option ? opt_option.threshold : DEFAULT_THRESHOLD;
```

Comment:
let's make the default the same to native impl, which is `[0]`.
we can have our default in a higher level.